File: app/src/main/java/com/hault/codex_java/di/DatabaseModule.java
package com.hault.codex_java.di;

import android.content.Context;
import androidx.room.Room;

import com.hault.codex_java.data.local.dao.ArcDao;
import com.hault.codex_java.data.local.dao.ChapterDao;
import com.hault.codex_java.data.local.dao.CharacterDao;
import com.hault.codex_java.data.local.dao.CrossRefDao;
import com.hault.codex_java.data.local.dao.EventDao;
import com.hault.codex_java.data.local.dao.LocationDao;
import com.hault.codex_java.data.local.dao.WorldDao;
import com.hault.codex_java.data.local.db.AppDatabase;
import com.hault.codex_java.data.repository.ArcRepository;
import com.hault.codex_java.data.repository.ChapterRepository;
import com.hault.codex_java.data.repository.CharacterRepository;
import com.hault.codex_java.data.repository.CrossRefRepository;
import com.hault.codex_java.data.repository.EventRepository;
import com.hault.codex_java.data.repository.LocationRepository;
import com.hault.codex_java.data.repository.WorldRepository;
import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.android.qualifiers.ApplicationContext;
import dagger.hilt.components.SingletonComponent;
import javax.inject.Singleton;


@Module
@InstallIn(SingletonComponent.class)
public class DatabaseModule {

    @Provides
    @Singleton
    public static AppDatabase provideAppDatabase(@ApplicationContext Context context) {
        return Room.databaseBuilder(
                context,
                AppDatabase.class,
                "codex_database"
        ).addMigrations(
                AppDatabase.MIGRATION_2_3,
                AppDatabase.MIGRATION_3_4,
                AppDatabase.MIGRATION_4_5
        ).build();
    }

    @Provides
    public static WorldDao provideWorldDao(AppDatabase appDatabase) {
        return appDatabase.worldDao();
    }

    @Provides
    public static CharacterDao provideCharacterDao(AppDatabase appDatabase) {
        return appDatabase.characterDao();
    }

    @Provides
    public static LocationDao provideLocationDao(AppDatabase appDatabase) {
        return appDatabase.locationDao();
    }

    @Provides
    public static EventDao provideEventDao(AppDatabase appDatabase) {
        return appDatabase.eventDao();
    }

    @Provides
    public static ArcDao provideArcDao(AppDatabase appDatabase) {
        return appDatabase.arcDao();
    }

    @Provides
    public static ChapterDao provideChapterDao(AppDatabase appDatabase) {
        return appDatabase.chapterDao();
    }

    @Provides
    public static CrossRefDao provideCrossRefDao(AppDatabase appDatabase) {
        return appDatabase.crossRefDao();
    }

    @Provides
    @Singleton
    public static WorldRepository provideWorldRepository(WorldDao worldDao) {
        return new WorldRepository(worldDao);
    }

    @Provides
    @Singleton
    public static CharacterRepository provideCharacterRepository(CharacterDao characterDao) {
        return new CharacterRepository(characterDao);
    }

    @Provides
    @Singleton
    public static LocationRepository provideLocationRepository(LocationDao locationDao) {
        return new LocationRepository(locationDao);
    }

    @Provides
    @Singleton
    public static EventRepository provideEventRepository(EventDao eventDao) {
        return new EventRepository(eventDao);
    }

    @Provides
    @Singleton
    public static ArcRepository provideArcRepository(ArcDao arcDao) {
        return new ArcRepository(arcDao);
    }

    @Provides
    @Singleton
    public static ChapterRepository provideChapterRepository(ChapterDao chapterDao) {
        return new ChapterRepository(chapterDao);
    }

    @Provides
    @Singleton
    public static CrossRefRepository provideCrossRefRepository(CrossRefDao crossRefDao) {
        return new CrossRefRepository(crossRefDao);
    }
}

File: app/src/main/java/com/hault/codex/di/DatabaseModule.kt
package com.hault.codex.di

import android.content.Context
import androidx.room.Room
import com.hault.codex.data.local.AppDatabase
import com.hault.codex.data.local.CharacterDao
import com.hault.codex.data.local.WorldDao
import com.hault.codex.data.local.LocationDao
import com.hault.codex.data.repository.CharacterRepository
import com.hault.codex.data.repository.LocationRepository
import com.hault.codex.data.local.EventDao
import com.hault.codex.data.repository.EventRepository
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.qualifiers.ApplicationContext
import dagger.hilt.components.SingletonComponent
import javax.inject.Singleton

import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase

@Module
@InstallIn(SingletonComponent::class)
object DatabaseModule {

    val MIGRATION_2_3 = object : Migration(2, 3) {
        override fun migrate(database: SupportSQLiteDatabase) {
            database.execSQL("CREATE TABLE IF NOT EXISTS events (id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, name TEXT NOT NULL, description TEXT, date TEXT NOT NULL, worldId INTEGER NOT NULL, FOREIGN KEY(worldId) REFERENCES worlds(id) ON UPDATE NO ACTION ON DELETE CASCADE)")
            database.execSQL("CREATE INDEX IF NOT EXISTS index_events_worldId ON events(worldId)")
        }
    }

    @Provides
    @Singleton
    fun provideAppDatabase(@ApplicationContext context: Context): AppDatabase {
        return Room.databaseBuilder(
            context,
            AppDatabase::class.java,
            "codex_database"
        ).addMigrations(MIGRATION_2_3).build()
    }

    @Provides
    fun provideWorldDao(appDatabase: AppDatabase): WorldDao {
        return appDatabase.worldDao()
    }

    @Provides
    fun provideCharacterDao(appDatabase: AppDatabase): CharacterDao {
        return appDatabase.characterDao()
    }

    @Provides
    fun provideLocationDao(appDatabase: AppDatabase): LocationDao {
        return appDatabase.locationDao()
    }

    @Provides
    fun provideEventDao(appDatabase: AppDatabase): EventDao {
        return appDatabase.eventDao()
    }

    @Provides
    @Singleton
    fun provideCharacterRepository(characterDao: CharacterDao): CharacterRepository {
        return CharacterRepository(characterDao)
    }

    @Provides
    @Singleton
    fun provideLocationRepository(locationDao: LocationDao): LocationRepository {
        return LocationRepository(locationDao)
    }

    @Provides
    @Singleton
    fun provideEventRepository(eventDao: EventDao): EventRepository {
        return EventRepository(eventDao)
    }
}
